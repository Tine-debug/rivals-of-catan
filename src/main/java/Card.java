// Card.java
// Quick & dirty, Basic-set only, refactored to reduce duplication

import java.io.IOException;

public class Card implements Comparable<Card> {

    // ---------- Public fields (keep simple for the take-home) ----------
    public String name, theme, type, cost, oneOf;
    public String cardText;

    public Points points;

    public String placement;

    private Logic logic;

    // Regions track “stored” resources by rotating; here we model it as an int
    // (0..3)
    public int regionProduction = 0;
    // Regions use production die faces (1..6). 0 means “not a region” / unassigned.
    public int diceRoll = 0;

    // ---------- Global piles for the Basic set ----------
    public static Cardstacks stacks = Cardstacks.getInstance();

    // ---------- Construction ----------
    public Card() {
    }

    public Card(String name, String theme, String type,
            String germanName, String placement,
            String oneOf, String cost,
            String victoryPoints, String CP, String SP, String FP,
            String PP, String LP, String KP, String Requires,
            String cardText, String protectionOrRemoval) {
        this.name = name;
        this.theme = theme;
        this.type = type;
        this.placement = placement;
        this.oneOf = oneOf;
        this.cost = cost;
        this.points = new Points(victoryPoints, CP, SP, FP, PP, LP, KP);
        this.cardText = cardText;

        this.logic = LogicFactory.createLogic(placement, name, type);
    }

    @Override
    public String toString() {
        return name;
    }

    @Override
    public int compareTo(Card o) {
        return this.name.compareToIgnoreCase(o.name);
    }

    public static void loadBasicCards(String jsonPath) throws IOException {
        Cardstacks.loadBasicCards(jsonPath);
    }

    public boolean applyEffect(Player active, Player other, int row, int col) {
        String nm = (this.name == null ? "" : this.name);
        System.out.println("ApplyEffect: " + nm + " at (" + row + "," + col + ")");
        return logic.applyEffect(active, other, row, col, this);
    }

    public String summarizePoints(){
        return points.summarizePoints();
    }

}
