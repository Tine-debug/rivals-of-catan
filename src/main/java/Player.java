
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.Set;

public class Player {

    public Points points = new Points("2", null, null, null, null, null, null);

    public int tradeRate = 3; 
    public boolean isBot = false;

    public Set<String> flags = new HashSet<>();

    public Map<String, Integer> resources = new HashMap<>();

    public List<Card> hand = new ArrayList<>();

    public Principality principality;

    public int lastSettlementRow = -1, lastSettlementCol = -1;

    private final Scanner in = new Scanner(System.in);

    public Player() {
        String[] all = { "Brick", "Grain", "Lumber", "Wool", "Ore", "Gold", "Any" };
        principality = new Principality();
        resources = new java.util.HashMap<>();
        for (String r : all)
            resources.put(r, 0);
    }

    public void sendMessage(Object m) {
        System.out.println(m);
    }

    public String receiveMessage() {
        System.out.print("> ");
        return in.nextLine();
    }

    public Card getCard(int r, int c) {
        return principality.getCard(r, c);
    }

    public void placeCard(int r, int c, Card card) {
        principality.placeCard(r, c, card);
    }


    public int expandAfterEdgeBuild(int col) {
        return principality.expandAfterEdgeBuild(col);
    }


    public boolean hasInPrincipality(String name) {
        return principality.hasInPrincipality(name);
    }

    public String printPrincipality() {
        return principality.printPrincipality(this);
    }


    public String printHand() {
        StringBuilder sb = new StringBuilder();
        sb.append("Hand (").append(hand.size()).append("):\n");
        for (int i = 0; i < hand.size(); i++) {
            Card c = hand.get(i);
            if (c == null)
                continue;
            String cost = (c.cost == null || c.cost.isBlank()) ? "-" : c.cost;
            String pts = c.summarizePoints();
            sb.append("  [").append(i).append("] ")
                    .append(c.name == null ? "Unknown" : c.name)
                    .append("   {cost: ").append(cost).append("} ")
                    .append(pts.isEmpty() ? "" : pts)
                    .append("\n").append(c.cardText == null ? "" : "\t" + c.cardText + "\n");
        }
        return sb.toString();
    }

  
    public boolean hasTradeTokenAgainst(Player opp) {
        if (this.points.commercePoints < 3) return false;
        if (opp == null) return true;
        return (this.points.commercePoints > opp.points.commercePoints);
    }

    public boolean hasStrengthTokenAgainst(Player opp) {
        if (this.points.strengthPoints < 3) return false;
        if (opp == null) return true;
        return (this.points.strengthPoints > opp.points.strengthPoints);
    }

    // Final score used for win check: base VP + 1 per advantage token against
    // opponent
    public int currentScoreAgainst(Player opp) {
        int score = this.points.victoryPoints;
        if (hasTradeTokenAgainst(opp))
            score += 1;
        if (hasStrengthTokenAgainst(opp))
            score += 1;
        return score;
    }


    // ------------- Resources (per-region, not pooled) -------------

    // Map a resource name to its Region card name
    private String resourceToRegion(String type) {
        if (type == null)
            return null;
        String t = type.trim().toLowerCase();
        switch (t) {
            case "brick":
                return "Hill";
            case "grain":
                return "Field";
            case "lumber":
                return "Forest";
            case "wool":
                return "Pasture";
            case "ore":
                return "Mountain";
            case "gold":
                return "Gold Field";
            case "any":
                return "Any";
            default:
                return null;
        }
    }



    public int totalAllResources() {
        return principality.totalAllResources();
    }

    // Count stored resources of a specific resource type across the board
    public int getResourceCount(String type) {
        return principality.getResourceCount(type);
    }

    // Gain 1 resource of a type: add to the matching region with the LOWEST stock
    // (<3)
    // If "Any", ask the player which resource to take.
    public void gainResource(String type) {
        String t = type;
        if (t == null || t.equalsIgnoreCase("Any")) {
            sendMessage("PROMPT: Choose resource to gain (Brick/Grain/Lumber/Wool/Ore/Gold):");
            t = receiveMessage();
        }
        String regionName = resourceToRegion(t);
        if (regionName == null || "Any".equals(regionName)) {
            sendMessage("Unknown resource '" + t + "'. Ignored.");
            return;
        }

        java.util.List<Card> regs = principality.findRegions(regionName);
        if (regs.isEmpty()) {
            sendMessage("No region for resource " + t + " is present.");
            return;
        }

        Card best = null;
        int bestVal = Integer.MAX_VALUE;
        for (Card r : regs) {
            int v = Math.max(0, Math.min(3, r.regionProduction));
            if (v < bestVal) {
                bestVal = v;
                best = r;
            }
        }
        if (best != null && best.regionProduction < 3) {
            best.regionProduction += 1;
        } else {
            sendMessage("No storage space on any " + regionName + " (already 3/3).");
        }
    }

 
    public boolean removeResource(String type, int n) {
        return principality.removeResource(type, n);
    }


    public void setResourceCount(String type, int n) {
        principality.setResourceCount(type, n);
    }

    // ------------- Hand -------------
    public int handSize() {
        return hand.size();
    }

    public void addToHand(Card c) {
        hand.add(c);
    }

    public Card removeFromHandByName(String nm) {
        for (int i = 0; i < hand.size(); i++) {
            Card c = hand.get(i);
            if (c != null && c.name != null && c.name.equalsIgnoreCase(nm)) {
                return hand.remove(i);
            }
        }
        return null;
    }

    // ------------- Cheap prompts used by Server -------------
    public String chooseResource() {
        sendMessage("PROMPT: Choose resource:");
        return receiveMessage();
    }
}