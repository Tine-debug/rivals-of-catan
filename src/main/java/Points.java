
public class Points {

    public int skillPoints = 0;
    public int strengthPoints = 0;
    public int victoryPoints = 0;
    public int progressPoints = 0;
    public int commercePoints = 0;
    public int sailPoints = 0;
    public int canonPoints = 0;

    public Points() {

    }

//String victoryPoints, String CP, String SP, String FP, String PP, String LP, String KP
    public Points(String victoryPoints, String CP, String SP, String FP, String PP, String LP, String KP) {
        this.victoryPoints = asInt(victoryPoints, 0);
        this.commercePoints = asInt(CP, 0);
        this.strengthPoints = asInt(SP, 0);
        this.skillPoints = asInt(FP, 0);
        this.sailPoints = asInt(LP, 0);
        this.canonPoints = asInt(KP, 0);

    }

    public static Points addPoints(Points points1, Points points2) {
        Points result = new Points();

        result.victoryPoints = points1.victoryPoints + points2.victoryPoints;
        result.commercePoints = points1.commercePoints + points2.commercePoints;
        result.strengthPoints = points1.strengthPoints + points2.strengthPoints;
        result.skillPoints = points1.skillPoints + points2.skillPoints;
        result.sailPoints = points1.sailPoints + points2.sailPoints;
        result.canonPoints = points1.canonPoints + points1.canonPoints;

        return result;

    }

    private int asInt(String s, int def) {
        try {
            if (s == null) {
                return def;
            }
            return Integer.parseInt(s.trim());
        } catch (Exception e) {
            return def;
        }
    }

    public String summarizePoints() {
        StringBuilder t = new StringBuilder();
        if (victoryPoints > 0 || commercePoints > 0 || progressPoints > 0 || strengthPoints > 0
                || skillPoints > 0) {
            t.append("[");
            if (victoryPoints > 0) {
                t.append("VP").append(victoryPoints).append(" ");
            }
            if (commercePoints > 0) {
                t.append("CP").append(commercePoints).append(" ");
            }
            if (strengthPoints > 0) {
                t.append("SP").append(strengthPoints).append(" ");
            }
            if (skillPoints > 0) {
                t.append("FP").append(skillPoints).append(" ");
            }
            if (progressPoints > 0) {
                t.append("PP").append(progressPoints).append(" ");
            }
            if (t.charAt(t.length() - 1) == ' ') {
                t.deleteCharAt(t.length() - 1);
            }
            t.append("]");
        }
        return t.toString();
    }
}
