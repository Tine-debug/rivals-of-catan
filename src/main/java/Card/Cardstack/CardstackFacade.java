package Card.Cardstack;

import Player.Player;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import Card.Card;

public class CardstackFacade {

    private final RegionManager regionManager;
    private final CenterBuildingManager buildingManager;
    private final EventManager eventManager;
    private final DrawStackManager drawStackManager;
    private final CardUtils cardUtils;

    private CardstackFacade() {
        this.regionManager = new RegionManager();
        this.buildingManager = new CenterBuildingManager();
        this.eventManager = new EventManager();
        this.drawStackManager = new DrawStackManager();
        this.cardUtils = new CardUtils();
    }

    static private CardstackFacade instance = null;

    static public CardstackFacade getInstance() {
        if (instance == null) {
            instance = new CardstackFacade();
        }
        return instance;
    }

    public void loadCards(String jsonPath, String theme) throws IOException{
        List<Card> all = LoadCards.loadCards(jsonPath, true, theme);
        int allsize = all.size();

        buildingManager.initializeCenterCards(all);
        regionManager.initializeRegions(all);
        eventManager.initializeEvents(all, true);
        drawStackManager.initializeStacks(all, true, allsize);

    }

    public List<Card> getStack(String Name) {
        return this.buildingManager.getStack(Name);
    }

    public List<Card> loadCardsForTesting(String jsonPath, String desiredTheme, boolean shuffle, boolean loadmulitpe) throws IOException {
        List<Card> all = LoadCards.loadCards(jsonPath, loadmulitpe, "basic");
        List<Card> out = new ArrayList<>(all);
        buildingManager.initializeCenterCards(all);
        regionManager.initializeRegions(all);
        eventManager.initializeEvents(all, shuffle);
        drawStackManager.initializeStacks(all, shuffle, out.size());
        return out;
    }

    public void initializePlayerPrincipality(Player p, int[][] dice, int center, int idx) {
        this.buildingManager.inizializePrincipiality(p, dice, center, idx, regionManager.getRegions());
    }

    public void shuffleRegions() {
        regionManager.shuffle();
    }

    public Card drawRegionCard() {
        return regionManager.draw();
    }

    public int getRegionStackSize() {
        return regionManager.size();
    }

    public String printRegionStack() {
        return regionManager.print();
    }

    public void resetEventStack() {
        eventManager.reset();
    }

    public Card drawEventCard() {
        return eventManager.draw();
    }

    public void placeCardBottomStack(Card c, int stackNum) {
        drawStackManager.placeAtBottom(c, stackNum);
    }

    public void drawCardFromStack(int stack, Player p) {
        drawStackManager.drawCard(stack, p);
    }

    public void chooseCardFromStack(int stack, Player p) {
        drawStackManager.chooseCard(stack, p);
    }

    public void drawCard(int stack, Player p) {
        drawStackManager.drawCard(stack, p);
    }

    public int[] placeCenterCard(Player active, Player other, String cardType) {
        return buildingManager.placeCenterCard(active, other, cardType);
    }

    public List<String> getCenterBuildingCosts() {
        return buildingManager.getBuildingCosts();
    }

    public Card pickRegionByNameOrIndex(String spec) {
        return regionManager.pickByNameOrIndex(spec);
    }

    public Card findUndicedRegionByName(String name) {
        return regionManager.findUndiced(name);
    }

    public Card popCardByName(List<Card> cards, String name) {
        return CardUtils.popCardByName(cards, name);
    }

}
