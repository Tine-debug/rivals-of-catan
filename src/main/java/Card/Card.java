package Card;

import java.io.IOException;
import Card.logic.*;
import Points.Points;
import Player.Player;

public class Card implements Comparable<Card> {

    // ---------- Public fields (keep simple for the take-home) ----------
    private String name, theme, type, cost;
    private boolean oneOf;
    private String cardText;

    private Points points;

    private String placement;

    private Logic logic;

    // Regions track “stored” resources by rotating; here we model it as an int
    // (0..3)
    private int regionProduction = 0;
    // Regions use production die faces (1..6). 0 means “not a region” / unassigned.
    private int diceRoll = 0;

    // ---------- Construction ----------
    public Card() {
    }

    protected Card(String name, String theme, String type,
        String placement, boolean oneOf, String cost,
        Points points, String cardText, Logic logic) {
        this.name = name;
        this.theme = theme;
        this.type = type;
        this.placement = placement;
        this.oneOf = oneOf;
        this.cost = cost;
        this.points = points;
        this.cardText = cardText;
        this.logic = logic;
    }


    //getters
    public String getTheme() {
        return theme;
    }

    public String getType() {
        return type;
    }

    public String getCost() {
        return cost;
    }

    public boolean getOneOf() {
        return oneOf;
    }

    public String getCardText() {
        return cardText;
    }

    public String getName() {
        return name;
    }

    public String getPlacement() {
        return placement;
    }

    public Points getPoints() {
        return points;
    }

    public int getRegionProduction() {
        return regionProduction;
    }

    public int getdiceRoll() {
        return diceRoll;
    }

    public void setPoints(Points newPoints) {
        points = newPoints;

    }

    public void setRegionProduction(int newcount) {
        if (newcount >= 0 && newcount <= 3) {
            regionProduction = newcount;
        }

    }

    public void setDiceRoll(int newdice) {
        if (newdice >= 0 && newdice <= 6) {
            diceRoll = newdice;
        }

    }

    @Override
    public String toString() {
        if (name != null) {
            return name;
        }
        return "?";
    }

    @Override
    public int compareTo(Card o) {
        return this.name.compareToIgnoreCase(o.name);
    }

    public boolean applyEffect(Player active, Player other, int row, int col) {
        String nm = (this.name == null ? "" : this.name);
        System.out.println("ApplyEffect: " + nm + " at (" + row + "," + col + ")");
        return logic.applyEffect(active, other, row, col, this);
    }

    public String summarizePoints() {
        return points.summarizePoints();
    }

}
