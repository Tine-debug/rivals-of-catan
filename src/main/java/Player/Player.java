package Player;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Scanner;
import java.util.Set;

import Points.*;
import Card.Card;

public class Player {



    private Points points;
    
    

    //private int tradeRate = 3;
    private boolean isBot = false;

    private Set<String> flags = new HashSet<>();

    private Map<String, Integer> resources = new HashMap<>();

    private final Hand hand = new Hand();

    private final Principality principality;

    private int lastSettlementRow = -1, lastSettlementCol = -1;

    private final Scanner in = new Scanner(System.in);

    public Player() {
        String[] all = {"Brick", "Grain", "Lumber", "Wool", "Ore", "Gold", "Any"};
        principality = new Principality();
        resources = new java.util.HashMap<>();
        for (String r : all) {
            resources.put(r, 0);
        }
        PointsBuilder pointsBuilder = new PointsBuilder();
        pointsBuilder.victoryPoints(2);
        this.points = pointsBuilder.build();
    }

    public void sendMessage(Object m) {
        System.out.println(m);
    }

    public String receiveMessage() {
        System.out.print("> ");
        return in.nextLine();
    }

    //getters
    public Set<String> getFlags(){
        return this.flags;
    }

    public Principality getPrincipality(){
        return principality;
    }

    public Hand getHand(){
        return hand;
    }

    public boolean getIsBot(){
        return isBot;
    }

    public int getLastSettlementRow(){
        return lastSettlementRow;
    }

    public int getLastSettlementCol(){
        return lastSettlementCol;
    }

    //setters
    public void setFlags(Set<String> newFlags){
        this.flags = newFlags;
    }

    public void setIsBot(boolean newvalue){
        this.isBot = newvalue;
    }

    public void setLastSettlementRow(int newvalue){
        this.lastSettlementRow = newvalue;
    }
    
    public void setLastSettlementCol(int newvalue){
        this.lastSettlementCol = newvalue;
    }

    

    //


    public void removeFlag(String toRemove){
        flags.remove(toRemove);
    }

    public Points getPoints(){
        return this.points;
    }

    public void setPoints(Points newValue){
        this.points = newValue;
    }



    public Card getCard(int r, int c) {
        return principality.getCard(r, c);
    }

    public void placeCard(int r, int c, Card card) {
        principality.placeCard(r, c, card);
    }

    public int expandAfterEdgeBuild(int col) {
        return principality.expandAfterEdgeBuild(col);
    }

    public boolean hasInPrincipality(String name) {
        return principality.hasInPrincipality(name);
    }

    public String printPrincipality() {
        return principality.printPrincipality(this);
    }

    public String printHand() {
        return hand.printHand();
    }

    public boolean hasTradeTokenAgainst(Player opp) {
        if (this.points.commercePoints < 3) {
            return false;
        }
        if (opp == null) {
            return true;
        }
        return (this.points.commercePoints > opp.points.commercePoints);
    }

    public boolean hasStrengthTokenAgainst(Player opp) {
        if (this.points.strengthPoints < 3) {
            return false;
        }
        if (opp == null) {
            return true;
        }
        return (this.points.strengthPoints > opp.points.strengthPoints);
    }

    // Final score used for win check: base VP + 1 per advantage token against
    // opponent
    public int currentScoreAgainst(Player opp) {
        int score = this.points.victoryPoints;
        if (hasTradeTokenAgainst(opp)) {
            score += 1;
        }
        if (hasStrengthTokenAgainst(opp)) {
            score += 1;
        }
        return score;
    }

    // ------------- Resources (per-region, not pooled) -------------
    // Map a resource name to its Region card name
    private String resourceToRegion(String type) {
        if (type == null) {
            return null;
        }
        String t = type.trim().toLowerCase();
        switch (t) {
            case "brick":
                return "Hill";
            case "grain":
                return "Field";
            case "lumber":
                return "Forest";
            case "wool":
                return "Pasture";
            case "ore":
                return "Mountain";
            case "gold":
                return "Gold Field";
            case "any":
                return "Any";
            default:
                return null;
        }
    }

    public int totalAllResources() {
        return principality.totalAllResources();
    }

    // Count stored resources of a specific resource type across the board
    public int getResourceCount(String type) {
        return principality.getResourceCount(type);
    }

    // Gain 1 resource of a type: add to the matching region with the LOWEST stock
    // (<3)
    // If "Any", ask the player which resource to take.
    public void gainResource(String type) {
        String t = type;
        if (t == null || t.equalsIgnoreCase("Any")) {
            sendMessage("PROMPT: Choose resource to gain (Brick/Grain/Lumber/Wool/Ore/Gold):");
            t = receiveMessage();
        }
        String regionName = resourceToRegion(t);
        if (regionName == null || "Any".equals(regionName)) {
            sendMessage("Unknown resource '" + t + "'. Ignored.");
            return;
        }

        java.util.List<Card> regs = principality.findRegions(regionName);
        if (regs.isEmpty()) {
            sendMessage("No region for resource " + t + " is present.");
            return;
        }

        Card best = null;
        int bestVal = Integer.MAX_VALUE;
        for (Card r : regs) {
            int v = Math.max(0, Math.min(3, r.getRegionProduction()));
            if (v < bestVal) {
                bestVal = v;
                best = r;
            }
        }
        if (best != null && best.getRegionProduction() < 3) {
            best.setRegionProduction(best.getRegionProduction() + 1);
        } else {
            sendMessage("No storage space on any " + regionName + " (already 3/3).");
        }
    }

    public boolean removeResource(String type, int n) {
        return principality.removeResource(type, n);
    }

    public void setResourceCount(String type, int n) {
        principality.setResourceCount(type, n);
    }

    // ------------- Hand -------------
    public int handSize() {
        return hand.handSize();
    }

    public void addToHand(Card c) {
        hand.addToHand(c);
    }

    public Card removeFromHandByName(String nm) {
        return hand.removeFromHandByName(nm);
    }

    // ------------- Cheap prompts used by Server -------------
    public String chooseResource() {
        sendMessage("PROMPT: Choose resource:");
        return receiveMessage();
    }

    public void refundCost(String cost) {
        if (cost == null || cost.isBlank()) {
            return;
        }
        Map<String, Integer> need = parseCost(cost);
        for (var e : need.entrySet()) {
            setResourceCount(e.getKey(), getResourceCount(e.getKey()) + e.getValue());
        }
    }

    public Map<String, Integer> parseCost(String cost) {
        Map<String, Integer> m = new HashMap<>();
        if (cost == null) {
            return m;
        }

        // Accept strings like "LW", "AA", with optional spaces or separators ("L,W", "A
        // A")
        for (int i = 0; i < cost.length(); i++) {
            char ch = cost.charAt(i);
            if (Character.isWhitespace(ch) || ch == ',' || ch == ';' || ch == '+') {
                continue;
            }
            String res = letterToResource(ch);
            if (res != null) {
                m.put(res, m.getOrDefault(res, 0) + 1);
            }
            // else: silently ignore unknown chars
        }
        return m;
    }

    public boolean payCost(String cost) {
        if (cost == null || cost.isBlank()) {
            return true;
        }
        // Cost like "1 Brick, 1 Grain, 1 Wool, 1 Lumber" etc.
        Map<String, Integer> need = parseCost(cost);
        for (var e : need.entrySet()) {
            if (getResourceCount(e.getKey()) < e.getValue()) {
                return false;
            }
        }
        for (var e : need.entrySet()) {
            removeResource(e.getKey(), e.getValue());
        }
        return true;
    }

    private String letterToResource(char ch) {
        switch (Character.toUpperCase(ch)) {
            case 'B':
                return "Brick";
            case 'G':
                return "Grain";
            case 'L':
                return "Lumber";
            case 'W':
                return "Wool";
            case 'O':
                return "Ore";
            case 'A':
                return "Gold";
            default:
                return null; // unknown / ignore
        }
    }


    public boolean hasStrengthAdvantage(Player other) {
        return this.points.strengthPoints >= 3 && this.points.strengthPoints > other.points.strengthPoints;
    }

}
