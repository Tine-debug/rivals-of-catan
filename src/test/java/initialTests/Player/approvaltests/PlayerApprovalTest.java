import java.util.Vector;

import Card.*;
import Player.*;
import Turns.*;
import Points.*;
import Card.Cardstack.CardstackFacade;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import java.util.ArrayList;
import java.util.List;

public class PlayerApprovalTest {


    Player player;
    private static CardstackFacade stacks = CardstackFacade.getInstance();

    @BeforeAll
    static void setupall() {
        try {
            List<Card> BasicCards = stacks.loadCardsForTesting("cards.json", "basic", false, true);
        } catch (Exception e) {
        }
            System.err.println("Failed to load: ");
        }

    @BeforeEach
    void setupeach() {
        player = new Player();
    }

    @Test
    void emptyPrincipality() {
        Approvals.verify(player.printPrincipality());
    }

    @Test
    void addCardToPrincipality() {
        Cardbuilder cardbuilder = new Cardbuilder();
        cardbuilder.name("Hill");
        cardbuilder.type("Region");
        cardbuilder.placement("Center Card");
        Card hill = cardbuilder.build();
        hill.setRegionProduction(1);
        player.placeCard(0, 0, hill);
        Approvals.verify(player.printPrincipality());
    }


/*
    @Test
    void addMultipleCardsAndHand() {
        Card forest = new Card("Forest", "Region", null);
        forest.setRegionProduction(2);
        Card settlement = new Card("Settlement", "Road", null);
        //String victoryPoints, String CP, String SP, String FP, String PP, String LP, String KP
        settlement.setPoints(new Points("2", null, null, null, null, null, null));        

        player.placeCard(0, 0, forest);
        player.placeCard(1, 1, settlement);

        Card cardInHand = new Card("Mill", "Building", "2 Grain");
        cardInHand.setPoints(new Points(null, null, "1", null, null, null, null));
        player.addToHand(cardInHand);

        Approvals.verify(player.printPrincipality() + "\n" + player.printHand());
    }
*/

    @Test
    void addMultipleCardsAndHand() {
        Cardbuilder cardbuilder = new Cardbuilder();


        cardbuilder.name("Forest");
        cardbuilder.type("Region");
        Card forest = cardbuilder.build();
        forest.setRegionProduction(2);

        cardbuilder = new Cardbuilder();
        cardbuilder.name("Settlement");
        cardbuilder.type("Road");
        PointsBuilder pointsBuilder = new PointsBuilder();
        pointsBuilder.victoryPoints(2);
        cardbuilder.points(pointsBuilder.build());
        Card settlement = cardbuilder.build();
        player.placeCard(0, 0, forest);
        player.placeCard(1, 1, settlement);


        cardbuilder = new Cardbuilder();
        pointsBuilder = new PointsBuilder();
        cardbuilder.name("Mill");
        cardbuilder.type("Building");
        cardbuilder.cost("2 Grain");

        pointsBuilder.strengthPoints(1);
        cardbuilder.points(pointsBuilder.build());
        Card cardInHand = cardbuilder.build();
        player.addToHand(cardInHand);

        Approvals.verify(player.printPrincipality() + "\n" + player.printHand());
    }

    @Test
    void resourceManipulation() {
        Cardbuilder cardbuilder = new Cardbuilder();
        cardbuilder.name("Forest");
        cardbuilder.type("Region");
        Card forest1 = cardbuilder.build();
        forest1.setRegionProduction(0);
        Card forest2 = cardbuilder.build();
        forest2.setRegionProduction(1);

        player.placeCard(0,0,forest1);
        player.placeCard(0,1,forest2);

        player.gainResource("Lumber"); // adds 1 to lowest
        player.gainResource("Lumber"); // adds 1 to the same or next lowest
        player.removeResource("Lumber", 1); // removes highest
        player.setResourceCount("Lumber", 3); // redistribute to total 3
        player.setResourceCount("Lumber", 1); // reduce to 1

        Approvals.verify("Total Lumber: " + player.getResourceCount("Lumber"));
    }

    @Test
    void scoreAndTokens() {
        Player opp = new Player();
        player.getPoints().victoryPoints = 3;
        player.getPoints().commercePoints = 5;
        player.getPoints().strengthPoints = 2;

        opp.getPoints().victoryPoints = 2;
        opp.getPoints().commercePoints = 1;
        opp.getPoints().strengthPoints = 0;

        String scoreReport = "Player score vs opponent: " + player.currentScoreAgainst(opp) +
                "\nHas trade token: " + player.hasTradeTokenAgainst(opp) +
                "\nHas strength token: " + player.hasStrengthTokenAgainst(opp);

        Approvals.verify(scoreReport);
    }

    @Test
    void expandAfterEdgeBuild() {
        Cardbuilder cardbuilder = new Cardbuilder();
        cardbuilder.name("Forest");
        Card c = cardbuilder.build();
        player.placeCard(0,0,c);
        int newColLeft = player.expandAfterEdgeBuild(0); // triggers left expansion
        int newColRight = player.expandAfterEdgeBuild(player.getPrincipality().principality.get(0).size()-1); // triggers right expansion
        Approvals.verify("Left col now: " + newColLeft + ", Right col now: " + newColRight);
    }

    @Test
    void hasInPrincipalityCheck() {
        Cardbuilder cardbuilder = new Cardbuilder();
        cardbuilder.name("Mine");
        Card c = cardbuilder.build();
        player.placeCard(2,2,c);
        String report = "Has Mine: " + player.hasInPrincipality("Mine") +
                        ", Has Forest: " + player.hasInPrincipality("Forest");
        Approvals.verify(report);
    }

    @Test
    void removeFromHandByNameTest() {
        Cardbuilder cardbuilder = new Cardbuilder();
        cardbuilder.name("Card1");
        Card c1 = cardbuilder.build();
        cardbuilder.name("Card2");
        Card c2 = cardbuilder.build();
        player.addToHand(c1);
        player.addToHand(c2);

        Card removed = player.removeFromHandByName("Card1");
        Card notFound = player.removeFromHandByName("Unknown");

        Approvals.verify("Removed: " + removed.toString()+ ", NotFound: " + notFound);
    }

    @Test
    void gainAnyResourceWithPrompt() {
        Player promptPlayer = new Player() {
            @Override
            public String receiveMessage() {
                return "Brick"; // simulate choosing a resource
            }
        };
        Cardbuilder cardbuilder = new Cardbuilder();
        cardbuilder.name("Hill");
        cardbuilder.type("Region");
        Card hill = cardbuilder.build();
        hill.setRegionProduction(0);
        promptPlayer.placeCard(0,0,hill);

        promptPlayer.gainResource("Any");
        Approvals.verify("Brick count: " + promptPlayer.getResourceCount("Brick"));
    }
}
