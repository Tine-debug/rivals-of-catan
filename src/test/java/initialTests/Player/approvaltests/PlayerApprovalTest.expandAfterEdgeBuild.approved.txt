Left col now: 1, Right col now: 5